### PR TITLE
refactor(libsinsp): replace cyclewriter with sinsp_cycledumper

### PR DESCRIPTION
**Pull request by therealbobo**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

/kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

/area libsinsp

/area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

This substitute the cyclewriter with `sinsp_cycledumper`.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
action required: the dumper has been removed from the inspector (in `next()`). Now its logic has to be called explicitly in the main loop of the consumer.
```

### CODE REVIEW

File: userspace/libsinsp/sinsp_cycledumper.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+void sinsp_cycledumper::autodump_start(const std::string& dump_filename)
+{
+	if(!m_inspector)
+	{
+		throw sinsp_exception("inspector not opened yet");
+	}
+
+	m_dumper.reset(new sinsp_dumper);
```
Comment (incertum) [inline userspace/libsinsp/sinsp_cycledumper.cpp:106]: Can we consistently use smart pointers instead?

Comment (therealbobo) [reply userspace/libsinsp/sinsp_cycledumper.cpp:106]: Yeap! Absolutely! :)

File: userspace/libsinsp/sinsp_cycledumper.cpp [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+void sinsp_cycledumper::autodump_start(const std::string& dump_filename)
+{
+	if(!m_inspector)
+	{
+		throw sinsp_exception("inspector not opened yet");
+	}
+
+	m_dumper = std::make_unique<sinsp_dumper>();
```
Comment (incertum) [inline userspace/libsinsp/sinsp_cycledumper.cpp:106]: Could this be made idempotent?

File: userspace/libsinsp/sinsp_cycledumper.cpp [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+	m_event_count++;
+
+	if(m_has_started == false)
+	{
+		m_has_started = true;
+
+        if(m_duration_seconds > 0)
+        {
+            // timer setup
+            m_last_time = evt->get_ts() / 1000000000; // 10^(-9) because it's nanoseconds
```
Comment (incertum) [inline userspace/libsinsp/sinsp_cycledumper.cpp:212]: we should have some existing macros for time conversions.

File: userspace/libsinsp/test/savefile.ut.cpp [hunk 4]
```diff
@@ -43,20 +44,26 @@ TEST(savefile, filter)
 		sinsp inspector;
 		inspector.set_filter("proc.name=ifplugd");
 		inspector.open_savefile(RESOURCE_DIR "/sample.scap");
-		inspector.setup_cycle_writer(filtered_scap, 0, 0, 0, 0, true);
-		inspector.autodump_next_file();
+
+		sinsp_cycledumper* dumper = new sinsp_cycledumper(&inspector, filtered_scap,
```
Comment (incertum) [inline userspace/libsinsp/test/savefile.ut.cpp:48]: Also use smart pointer in the test, but actually can we make the test more realistic and call `sinsp_cycledumper::autodump_start` instead etc?

**Review by incertum — COMMENTED** [review]
[nit] some general indentation hiccups.

File: userspace/libsinsp/test/savefile.ut.cpp (hunk 4 above)
Comment (incertum) [reply userspace/libsinsp/test/savefile.ut.cpp]: Double-checking re making the test more realistic? Any updates on that?

Comment (therealbobo) [reply userspace/libsinsp/test/savefile.ut.cpp]: Didn’t forget about that! I’m looking into it! :)

File: userspace/libsinsp/sinsp_cycledumper.h [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    int m_rollover_mb; //!< Max scap file size in MB.
+    int m_duration_seconds; //!< Max duration for each capture in seconds.
+    int m_file_limit; //!< Max number of scap file generated.
+    unsigned long m_event_limit; //!< Max number of events for each catpure.
+    time_t m_last_time; //!< Last time of a capture.
+    int m_file_count_total; //!< Total number of files written.
+    int m_file_index; //!< Current file index.
+    bool m_has_started; //!< Indicates if the cycledumper has started for the first time.
+    unsigned long m_event_count; //!< Number of events of the current scap file.
+    std::string *m_past_names; //!< Ring buffer to maintain the file names for scap rotation.
```
Comment (incertum) [inline userspace/libsinsp/sinsp_cycledumper.h:73]: Maybe I am missing it, but where did the cleanup go from the old code base?

```
{
    if(m_past_names != NULL)
    {
	delete[] m_past_names;
    }
};
```

Comment (incertum) [reply userspace/libsinsp/sinsp_cycledumper.h:73]: Perhaps you have ideas to refactor that also, but not wanting to drag this PR out much longer. We can always continue working on it in follow up PRs.

**Review by incertum — COMMENTED** [review]
Nice job adding all new tests! Perhaps do a one more pass self-review as well and then we can merge it very soon!

**Review by incertum — DISMISSED** [review]
/approve

Thanks for this refactor and getting the cyclewriter out of sinsp.

**Review by jasondellaluce — APPROVED** [review]
/approve

**Review by incertum — APPROVED** [review]
/approve

**Pull request merged**